### ISSUE TITLE
[EXPL] [Arch] - evaluate transformer option (BERT)

### ISSUE THREAD

**Issue by Mars-Martiny** [Mars-Martiny/CITS5206_group1#60]

**Comment by aashutosh-chapagain** [comment]
@ChangLiu-doc  I have some ideas on this, so Im also assigning myself here. I will push some code into feature/bert-transformer branch and we can review changes over there.

**Comment by aashutosh-chapagain** [comment]
**Objective**
Create and evaluate a BERT-based transformer pipeline:

**Tasks:**
- Build a BERT-compatible dataloader
- Build a BERT classifier model
- Implement or integrate a BERT transformer
- Run comprehensive evaluations on the implemented model
- Fine-tune hyperparameters to identify the best performing configuration

**Acceptance Criteria:**
- Each component is independently tested and documented
- Evaluations compare performance across configurations
- Provide a summary of best options found with appropriate rationale

**Comment by aashutosh-chapagain** [comment]
Results:

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>run_name</th>
      <th>fine_tune</th>
      <th>pooling</th>
      <th>learning_rate</th>
      <th>dropout</th>
      <th>weight_decay</th>
      <th>class_weights</th>
      <th>accuracy</th>
      <th>f1_macro</th>
      <th>f1_weighted</th>
      <th>precision_macro</th>
      <th>recall_macro</th>
      <th>auto_classification_rate</th>
      <th>best_epoch</th>
      <th>best_val_metric</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>bert_energy_finetuned_cls_lr2e5_do03_wd001</td>
      <td>True</td>
      <td>cls</td>
      <td>0.00002</td>
      <td>0.3</td>
      <td>0.01</td>
      <td>False</td>
      <td>0.701681</td>
      <td>0.646702</td>
      <td>0.695754</td>
      <td>0.678235</td>
      <td>0.662766</td>
      <td>0.756303</td>
      <td>5</td>
      <td>0.606426</td>
    </tr>
    <tr>
      <th>4</th>
      <td>bert_energy_finetuned_cls_lr2e5_do03_wd001_cw</td>
      <td>True</td>
      <td>cls</td>
      <td>0.00002</td>
      <td>0.3</td>
      <td>0.01</td>
      <td>True</td>
      <td>0.705882</td>
      <td>0.630999</td>
      <td>0.705778</td>
      <td>0.649905</td>
      <td>0.652384</td>
      <td>0.542017</td>
      <td>4</td>
      <td>0.639734</td>
    </tr>
    <tr>
      <th>3</th>
      <td>bert_energy_finetuned_cls_lr2e5_do03_wd005</td>
      <td>True</td>
      <td>cls</td>
      <td>0.00002</td>
      <td>0.3</td>
      <td>0.05</td>
      <td>False</td>
      <td>0.676471</td>
      <td>0.622797</td>
      <td>0.666323</td>
      <td>0.680601</td>
      <td>0.627546</td>
      <td>0.731092</td>
      <td>5</td>
      <td>0.602005</td>
    </tr>
    <tr>
      <th>5</th>
      <td>bert_energy_finetuned_cls_lr1e5_do03_wd001_cw</td>
      <td>True</td>
      <td>cls</td>
      <td>0.00001</td>
      <td>0.3</td>
      <td>0.01</td>
      <td>True</td>
      <td>0.668067</td>
      <td>0.598408</td>
      <td>0.674983</td>
      <td>0.609185</td>
      <td>0.630480</td>
      <td>0.180672</td>
      <td>4</td>
      <td>0.582932</td>
    </tr>
    <tr>
      <th>2</th>
      <td>bert_energy_finetuned_cls_lr3e5_do02_wd001</td>
      <td>True</td>
      <td>cls</td>
      <td>0.00003</td>
      <td>0.2</td>
      <td>0.01</td>
      <td>False</td>
      <td>0.659664</td>
      <td>0.581386</td>
      <td>0.660488</td>
      <td>0.607545</td>
      <td>0.583806</td>
      <td>0.558824</td>
      <td>3</td>
      <td>0.602199</td>
    </tr>
    <tr>
      <th>0</th>
      <td>bert_energy_finetuned_cls_lr1e5_do02_wd001</td>
      <td>True</td>
      <td>cls</td>
      <td>0.00001</td>
      <td>0.2</td>
      <td>0.01</td>
      <td>False</td>
      <td>0.638655</td>
      <td>0.510919</td>
      <td>0.628096</td>
      <td>0.553612</td>
      <td>0.504540</td>
      <td>0.516807</td>
      <td>5</td>
      <td>0.544092</td>
    </tr>
  </tbody>
</table>
</div>


<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>run_name</th>
      <th>learning_rate</th>
      <th>dropout</th>
      <th>weight_decay</th>
      <th>class_weights</th>
      <th>max_length</th>
      <th>threshold</th>
      <th>accuracy</th>
      <th>f1_macro</th>
      <th>f1_weighted</th>
      <th>precision_macro</th>
      <th>recall_macro</th>
      <th>auto_classification_rate</th>
      <th>meets_requirement</th>
      <th>best_epoch</th>
      <th>best_val_metric</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>bert_energy_BALANCED_cw_threshold07</td>
      <td>0.000020</td>
      <td>0.30</td>
      <td>0.01</td>
      <td>True</td>
      <td>160</td>
      <td>0.7</td>
      <td>0.697479</td>
      <td>0.628501</td>
      <td>0.699143</td>
      <td>0.638543</td>
      <td>0.661490</td>
      <td>0.714286</td>
      <td>True</td>
      <td>5</td>
      <td>0.614548</td>
    </tr>
    <tr>
      <th>1</th>
      <td>bert_energy_REGULARISED_do035</td>
      <td>0.000020</td>
      <td>0.35</td>
      <td>0.01</td>
      <td>False</td>
      <td>160</td>
      <td>0.8</td>
      <td>0.684874</td>
      <td>0.615872</td>
      <td>0.674860</td>
      <td>0.706170</td>
      <td>0.609887</td>
      <td>0.718487</td>
      <td>True</td>
      <td>5</td>
      <td>0.597454</td>
    </tr>
    <tr>
      <th>0</th>
      <td>bert_energy_BEST_lr25e6_do03</td>
      <td>0.000025</td>
      <td>0.30</td>
      <td>0.01</td>
      <td>False</td>
      <td>160</td>
      <td>0.8</td>
      <td>0.647059</td>
      <td>0.561351</td>
      <td>0.631333</td>
      <td>0.688634</td>
      <td>0.584335</td>
      <td>0.705882</td>
      <td>True</td>
      <td>4</td>
      <td>0.585802</td>
    </tr>
  </tbody>
</table>
</div>

So far, Best Region Around:

```
{
    "fine_tune": True,
    "pooling": "cls",
    "learning_rate": 2e-5,
    "dropout": 0.3,
    "weight_decay": 0.01,
    "max_length": 160
}
```

**Comment by aashutosh-chapagain** [comment]
I will work on integrating SafetyBERT pre-trained model to check if that brings any improvements.

**Comment by aashutosh-chapagain** [comment]
Completed on #100
